Handle Pyodide and pyodide-build unvendoring, and update Pyodide CI job

agriyakhetarpal opened the pull request:
### Description

Closes https://github.com/pyodide/pyodide#5585

This PR refactors the `update_pyodide_versions()` function in `tooling/src/hypothesistooling/__main__.py` which generates the relevant versions for Pyodide and its build tooling in the `test-pyodide:` CI job. The previous logic did not account for the fact that we had unvendored `pyodide-build` and moved it to a separate package that is released separately of the Pyodide runtime during the Pyodide 0.27 release cycle.

### Summary

With these changes, the logic per the periodic updates will:
- use the PyPI Simple API to find out the latest stable version of `pyodide-build`, like before
- scan [the Pyodide cross-build environments metadata file](https://github.com/pyodide/pyodide/blob/d878eb2059b877fd4d25623d162c73872cdb4530/pyodide-cross-build-environments.json), and select a stable version for the Pyodide runtime that satisfies the minimum and maximum `pyodide-build` version constraints
- always use this identified Pyodide version through the `pyodide xbuildenv install` command

Other changes I've made:
- Dropped the Emscripten version installation step as it's not needed for pure Python packages
- Removed a redundant step through which we were building wheels twice
- Upgraded to a newer Node.js version (22) over the current 18

### Ancillary notes

- [x] I have gone through the [Copyright and Licensing section](https://github.com/HypothesisWorks/hypothesis/blob/6c024aaf0dc8bc13a63eade36891c41af2f1e038/CONTRIBUTING.rst#copyright-and-licensing), and I confirm that I am not performing this change on work time.
- [x] I have added my name to the AUTHORS.rst file.

Zac-HD commented on tooling/src/hypothesistooling/__main__.py:
```diff
@@ -404,28 +404,88 @@ def update_django_versions():
-            python_version, emscripten_version = match.groups()
-            break
+
+    pyodide_release = sorted(
+        compatible_releases,
+        key=lambda release: tuple(int(x) for x in release["version"].split(".")),
+        reverse=True,
+    )[0]
```
Looks like you want `max(compatible_releases, key=...)`!  (and similarly above)

Zac-HD commented on tooling/src/hypothesistooling/__main__.py:
```diff
@@ -404,28 +404,88 @@ def update_django_versions():
+    stable_releases = [
+        release
+        for release in cross_build_environments_data["releases"].values()
+        if "a" not in release["version"]
+        and "b" not in release["version"]
+        and "rc" not in release["version"]
```
`if re.fullmatch(vers_re, release["version"])`?

Zac-HD commented on tooling/src/hypothesistooling/__main__.py:
```diff
@@ -404,28 +404,88 @@ def update_django_versions():
+        min_build_version = release.get("min_pyodide_build_version", "0.0.0")
+        max_build_version = release.get("max_pyodide_build_version", "999.999.999")
+
+        min_parts = [int(x) for x in min_build_version.split(".")]
+        max_parts = [int(x) for x in max_build_version.split(".")]
+
```
let's pull out a helper function:

```python
def version_tuple(v: str) -> tuple[int, int, int]:
    return tuple(int(x) for x in v.split("."))  # type: ignore
```

and then all of this logic can lean on tuple-comparison semantics, and thus:

```python
if (
    version_tuple(min_build_version)
    <= version_tuple(pyodide_build_version)
    <= version_tuple(max_build_version
):
    compatible_releases.append(release)
```

Zac-HD reviewed (commented):
Thanks, @agriyakhetarpal!  It looks like this should really stabilize our pyodide testing.

I've commented below assuming that you'll want to finish this up; we're also happy to take it from here if you'd prefer to hand it off.  Either way, thanks so much!

agriyakhetarpal commented:
Appreciate the feedback, @Zac-HD! I addressed your review comments. Yes, I intend to complete this. I marked it as a draft to see if the Pyodide testing would pass with the updated versions, or if there would be more work to do. It passed in the previous commits: https://github.com/HypothesisWorks/hypothesis/actions/runs/14348638355/job/40223061080?pr=4352, so I'll wrap this up, update the PR description, and mark it as ready for review.

agriyakhetarpal commented on .github/workflows/main.yml:
```diff
@@ -223,17 +223,17 @@ jobs:
-      # Note that the versions below must be updated in sync; we've automated
-      # that with `update_pyodide_versions()` in our weekly cronjob.
-      PYODIDE_VERSION: 0.27.3
+      NODE_VERSION: 22
```
I updated to a newer Node.js version, as we will be bumping the minimum away from 18 sometime soon. Our general recommendation is >=20.

agriyakhetarpal commented on AUTHORS.rst:
```diff
@@ -14,6 +14,7 @@ their individual contributions.
 * `Adam Sven Johnson <https://www.github.com/pkqk>`_
 * `Afrida Tabassum <https://github.com/oxfordhalfblood>`_ (afrida@gmail.com)
 * `Afonso Silva <https://github.com/ajcerejeira>`_ (ajcerejeira@gmail.com)
+* `Agrya Khetarpal <https://github.com/agriyakhetarpal>`_
```
```suggestion
* `Agriya Khetarpal <https://github.com/agriyakhetarpal>`_
```

Fixing my name's spelling, oops 😅

agriyakhetarpal commented on .github/workflows/main.yml:
```diff
@@ -223,17 +223,17 @@ jobs:
+      PYODIDE_VERSION: 0.27.5
+      PYODIDE_BUILD_VERSION: 0.30.0
       PYTHON_VERSION: 3.12.7
-      EMSCRIPTEN_VERSION: 3.1.58
```
From what I understand is that Hypothesis is a pure Python package, and therefore it does not need the Emscripten version or involve any compilation steps as a part of its wheel builds.

agriyakhetarpal commented on .github/workflows/main.yml:
```diff
@@ -246,17 +246,10 @@ jobs:
-      - name: Build
+      - name: Install pyodide-build and Pyodide cross-build environment
         run: |
-          # TODO remove https://github.com/pyodide/pyodide/issues/5585
-          pip install -U wheel==0.45.1
-          pip install pyodide-build==$PYODIDE_VERSION
-          cd hypothesis-python/
-          CFLAGS=-g2 LDFLAGS=-g2 pyodide build
```
The `CFLAGS=-g2 LDFLAGS=-g2 pyodide build` step is redundant, as:
1. We delegate `pyodide build` to `pypa/build` for building pure Python wheels, and the compilation flags won't do anything
2. We've been running `python -m build --wheel hypothesis-python --outdir dist/` anyway, which has been the wheel built by `pyodide build` in the previous

For pure Python packages, `pyodide build` does not provide any special provisions or invoke Pyodide-specific behaviour – using `pypa/build` is better.

agriyakhetarpal commented on .github/workflows/main.yml:
```diff
@@ -246,17 +246,10 @@ jobs:
         uses: actions/setup-node@b39b52d1213e96004bfcb1c61a8a6fa8ab84f3e8 # v4.0.1
         with:
           node-version: ${{ env.NODE_VERSION }}
-      - name: Set up Emscripten
-        uses: mymindstorm/setup-emsdk@6ab9eb1bda2574c4ddb79809fc9247783eaf9021 # v14
-        with:
-          version: ${{ env.EMSCRIPTEN_VERSION }}
```
Removing the Emscripten setup shaves off ~35 seconds from the CI job timings, and will reduce it more if the caches are evicted/invalidated. It isn't a lot, but I hope it's a welcome change. :)

agriyakhetarpal commented on tooling/src/hypothesistooling/__main__.py:
```diff
@@ -403,31 +403,65 @@ def update_django_versions():
 
 def update_pyodide_versions():
+
+    def version_tuple(v: str) -> tuple[int, int, int]:
+        return tuple(int(x) for x in v.split("."))  # type: ignore
```
I kept this as a nested function as it's only a one-liner; happy to move it outside as a private function if you suggest.

agriyakhetarpal reviewed (commented):
Here's a self-review that explains some of the changes I've made.

Zac-HD reviewed (approved):
Thanks again @agriyakhetarpal!
